autogpt_platform/backend/backend/integrations/oauth/google.py

```diff
@@ -35,36 +46,78 @@ def get_login_url(self, scopes: list[str], state: str) -> str:
         )
         return authorization_url
 
-    def exchange_code_for_tokens(self, code: str) -> OAuth2Credentials:
-        flow = self._setup_oauth_flow(None)
+    def exchange_code_for_tokens(
+        self, code: str, scopes: list[str]
+    ) -> OAuth2Credentials:
+        logger.debug(f"Exchanging code for tokens with scopes: {scopes}")
+
+        # Use the scopes from the initial request
+        flow = self._setup_oauth_flow(scopes)
         flow.redirect_uri = self.redirect_uri
-        flow.fetch_token(code=code)
 
-        google_creds = flow.credentials
-        username = self._request_email(google_creds)
+        logger.debug("Fetching token from Google")
 
-        # Google's OAuth library is poorly typed so we need some of these:
-        assert google_creds.token
-        assert google_creds.refresh_token
-        assert google_creds.expiry
-        assert google_creds.scopes
-        return OAuth2Credentials(
-            provider=self.PROVIDER_NAME,
-            title=None,
-            username=username,
-            access_token=SecretStr(google_creds.token),
-            refresh_token=SecretStr(google_creds.refresh_token),
-            access_token_expires_at=int(google_creds.expiry.timestamp()),
-            refresh_token_expires_at=None,
-            scopes=google_creds.scopes,
-        )
+        try:
+            # Disable scope check in fetch_token
+            flow.oauth2session.scope = None
+            token = flow.fetch_token(code=code)
+            logger.debug("Token fetched successfully")
+
+            # Get the actual scopes granted by Google
+            granted_scopes: list[str] = s.split() if (s := token.get("scope")) else []
+
+            logger.debug(f"Scopes granted by Google: {granted_scopes}")
+
+            google_creds = flow.credentials
+            logger.debug(f"Received credentials: {google_creds}")
+
+            logger.debug("Requesting user email")
+            username = self._request_email(google_creds)
+            logger.debug(f"User email retrieved: {username}")
+
+            if not google_creds.token:
+                raise ValueError("No access token received")
+            if not google_creds.refresh_token:
+                raise ValueError("No refresh token received")
+            if not google_creds.expiry:
+                raise ValueError("No expiry time received")
+            if not granted_scopes:
+                raise ValueError("No scopes received")
```

Comment:
Does any of these ever happen?

Reply:
Not yet

Reply:
Then for compactness (and not to raise false suspicions) I'd just keep the assert statements. Now it looks like there is a reasonably expected case where any or all of these attributes are not present, which isn't true.